core/src/main/scala/laserdisc/protocol/RESP.scala

```diff
@@ -393,75 +395,139 @@ sealed trait RESPFunctions { this: RESPCodecs =>
 
   final def stateOf: BitVector => String | BitVectorState =
     bits => bits.consume(BitsInByte) {
-      case `plus`   => Right(SimpleString)
-      case `minus`  => Right(Error)
-      case `colon`  => Right(Integer)
-      case `dollar` => Right(BulkString)
-      case `star`   => Right(RedisArray)
-      case other    => Left(s"unidentified RESP type (Hex: ${other.toHex})")
+      case `plus`
+           | `minus`
+           | `colon` => Right(NoSize)
+      case `dollar`  => Right(BulkSize)
+      case `star`    => Right(CollectionSize)
+      case other     => Left(s"unidentified RESP type when checking the state: unexpected value ${other.print} (Hex: ${other.toHex})")
     }
     .flatMap {
-      case (remainder, BulkString) =>
+      case (remainder, BulkSize) =>
         evalWithSizeDecodedFrom(remainder) {
-          case Left(_)   => IncompleteVector
+          case Left(_)   => Incomplete
           case Right(ds) =>
             val expectedBulkSize = (ds.value * BitsInByte) + crlf.size
 
             if (ds.value >= 0 && ds.remainder.size == expectedBulkSize)
-              CompleteVector // Complete: expected size
+              Complete
 
             else if (ds.value >= 0 && ds.remainder.size > expectedBulkSize) {
               val completeBulkSize = bits.bytes.indexOfSlice(crlfBytes, 0L) * BitsInByte + crlf.size + expectedBulkSize
               CompleteWithRemainder(bits.take(completeBulkSize), bits.drop(completeBulkSize)) // Complete plus some remainder
             }
 
             else if (ds.value == -1 && ds.remainder.size == 0)
-              CompleteVector // Complete: empty bulk
+              Complete
 
             else if (ds.value == -1 && ds.remainder.size > 0) {
               val completeBulkSize = 3 * BitsInByte + crlf.size
               CompleteWithRemainder(bits.take(completeBulkSize), bits.drop(completeBulkSize)) // Complete plus some remainder
             }
 
-            else
-              MissingBits(expectedBulkSize - ds.remainder.size)
+            else MissingBits(expectedBulkSize - ds.remainder.size)
         }
 
-      case (remainder, RedisArray) =>
-        arrayCodec.decode(remainder).fold(
-          _ => Right(IncompleteVector),
-          r => Right(CompleteAndDecoded(r))
-        )
+      case (remainder, CollectionSize) =>
+        failingEvalWithSizeDecodedFrom(remainder) {
+          case Left(_)  =>
+            Right(Incomplete)
+
+          case Right(ds) =>
```

Comment:
That way it could be easier to use one of the remainder instead of the other. Instead now the idea I used is that inside the cases I use the dot notation if it has to come from the result of the eval.

Reply:
replacing `consume` with `consumeThen` the situation improved a lot. I changed this as well as you suggest.